File: python/src/dexcost/adapters/aws_lambda.py
"""AWS Lambda cost adapter — compute cost from duration, memory, and region.

Pure function ``lambda_cost`` returns a dict with ``cost_usd`` (Decimal) and
``details`` breakdown.  Uses bundled pricing JSON; no network I/O.

Implements US-043.
"""

from __future__ import annotations

import json
import logging
from decimal import Decimal
from importlib import resources
from typing import Any

_log = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Bundled pricing data (loaded once at import time)
# ---------------------------------------------------------------------------

_pricing_data: dict[str, Any] | None = None


def _load_pricing() -> dict[str, Any]:
    """Load and cache the bundled AWS Lambda pricing JSON."""
    global _pricing_data
    if _pricing_data is not None:
        return _pricing_data

    ref = (
        resources.files("dexcost")
        .joinpath("adapters")
        .joinpath("data")
        .joinpath("aws_lambda_pricing.json")
    )
    raw = ref.read_text(encoding="utf-8")
    try:
        _pricing_data = json.loads(raw)
    except json.JSONDecodeError:
        _log.warning("Failed to parse AWS pricing data")
        _pricing_data = {}
    return _pricing_data


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def get_supported_regions() -> list[str]:
    """Return a sorted list of AWS region codes with bundled pricing data."""
    data = _load_pricing()
    return sorted(data["regions"].keys())


def lambda_cost(
    duration_ms: int,
    memory_mb: int,
    region: str,
) -> dict[str, Any]:
    """Calculate the cost of a single AWS Lambda invocation.

    This is a **pure function** — no I/O, no side effects.  It uses the
    bundled ``aws_lambda_pricing.json`` for rates.

    Args:
        duration_ms: Execution duration in milliseconds (>= 0).
        memory_mb: Allocated memory in MB (> 0).
        region: AWS region code (e.g. ``"us-east-1"``).

    Returns:
        Dict with:
        - ``cost_usd``: Total cost as a :class:`~decimal.Decimal`.
        - ``details``: Breakdown dict with ``region``, ``duration_ms``,
          ``memory_mb``, ``gb_seconds``, ``duration_cost_usd``,
          ``request_cost_usd``, ``rate_per_gb_second``.

    Raises:
        ValueError: If ``region`` is unknown, ``duration_ms`` < 0,
            or ``memory_mb`` <= 0.
    """
    # --- Input validation ---
    if duration_ms < 0:
        raise ValueError(f"duration_ms must be >= 0, got {duration_ms}")
    if memory_mb <= 0:
        raise ValueError(f"memory_mb must be > 0, got {memory_mb}")

    data = _load_pricing()
    region_pricing = data["regions"].get(region)
    if region_pricing is None:
        supported = ", ".join(sorted(data["regions"].keys()))
        raise ValueError(
            f"Unknown AWS region '{region}'. Supported regions: {supported}"
        )

    # --- Compute GB-seconds ---
    duration_seconds = Decimal(str(duration_ms)) / Decimal("1000")
    memory_gb = Decimal(str(memory_mb)) / Decimal("1024")
    gb_seconds = duration_seconds * memory_gb

    # --- Look up rates ---
    rate_per_gb_second = Decimal(region_pricing["duration_per_gb_second"])
    request_charge = Decimal(region_pricing["request_per_invocation"])

    # --- Calculate costs ---
    duration_cost = gb_seconds * rate_per_gb_second
    total_cost = duration_cost + request_charge

    return {
        "cost_usd": total_cost,
        "details": {
            "region": region,
            "duration_ms": duration_ms,
            "memory_mb": memory_mb,
            "gb_seconds": gb_seconds,
            "duration_cost_usd": duration_cost,
            "request_cost_usd": request_charge,
            "rate_per_gb_second": rate_per_gb_second,
        },
    }

File: typescript/src/adapters/aws-lambda.ts
/**
 * AWS Lambda cost adapter — compute cost from duration, memory, and region.
 *
 * `lambdaCost` is a pure function that returns the cost (in USD) and a
 * breakdown for a single Lambda invocation, using bundled pricing data
 * (no network I/O). Mirrors the Python SDK's `adapters/aws_lambda.py`.
 */

// Sprint 3 Theme E / §4.2.3 — Node 18 compat: runtime JSON load.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const _thisDir = dirname(fileURLToPath(import.meta.url));
const pricingData = JSON.parse(
  readFileSync(join(_thisDir, "data", "aws_lambda_pricing.json"), "utf-8"),
);

interface RegionPricing {
  duration_per_gb_second: Record<LambdaArchitecture, string>;
  request_per_invocation: string;
}

export type LambdaArchitecture = "x86_64" | "arm64";

interface LambdaPricingData {
  regions: Record<string, RegionPricing>;
}

const _pricing = pricingData as unknown as LambdaPricingData;

/** Breakdown of a Lambda invocation's cost. */
export interface LambdaCostDetails {
  region: string;
  architecture: LambdaArchitecture;
  durationMs: number;
  memoryMb: number;
  gbSeconds: number;
  durationCostUsd: number;
  requestCostUsd: number;
  ratePerGbSecond: number;
}

/** Result of {@link lambdaCost}. */
export interface LambdaCostResult {
  /** Total invocation cost in USD. */
  costUsd: number;
  /** Cost breakdown. */
  details: LambdaCostDetails;
}

/**
 * Return a sorted list of AWS region codes with bundled pricing data.
 */
export function getSupportedRegions(): string[] {
  return Object.keys(_pricing.regions).sort();
}

/**
 * Calculate the cost of a single AWS Lambda invocation.
 *
 * Pure function — no I/O, no side effects. Uses the bundled
 * `aws_lambda_pricing.json` for rates.
 *
 * @param durationMs - Execution duration in milliseconds (>= 0).
 * @param memoryMb - Allocated memory in MB (> 0).
 * @param region - AWS region code (e.g. `"us-east-1"`).
 * @param architecture - Lambda architecture. Defaults to `"x86_64"`.
 * @throws Error when an input is invalid or the region is unknown.
 */
export function lambdaCost(
  durationMs: number,
  memoryMb: number,
  region: string,
  architecture: LambdaArchitecture = "x86_64",
): LambdaCostResult {
  if (durationMs < 0) {
    throw new Error(`durationMs must be >= 0, got ${durationMs}`);
  }
  if (memoryMb <= 0) {
    throw new Error(`memoryMb must be > 0, got ${memoryMb}`);
  }
  if (architecture !== "x86_64" && architecture !== "arm64") {
    throw new Error(`Unsupported Lambda architecture '${String(architecture)}'`);
  }

  const regionPricing = _pricing.regions[region];
  if (regionPricing === undefined) {
    const supported = getSupportedRegions().join(", ");
    throw new Error(
      `Unknown AWS region '${region}'. Supported regions: ${supported}`,
    );
  }

  // GB-seconds = duration (s) * memory (GB)
  const durationSeconds = durationMs / 1000;
  const memoryGb = memoryMb / 1024;
  const gbSeconds = durationSeconds * memoryGb;

  const ratePerGbSecond = Number(regionPricing.duration_per_gb_second[architecture]);
  const requestCharge = Number(regionPricing.request_per_invocation);

  const durationCost = gbSeconds * ratePerGbSecond;
  const totalCost = durationCost + requestCharge;

  return {
    costUsd: totalCost,
    details: {
      region,
      architecture,
      durationMs,
      memoryMb,
      gbSeconds,
      durationCostUsd: durationCost,
      requestCostUsd: requestCharge,
      ratePerGbSecond,
    },
  };
}
